processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java

```diff
@@ -137,14 +137,12 @@ public BeanMappingMethod build() {
                 return null;
             }
 
-            if ( !method.getMappingOptions().isRestrictToDefinedMappings() ) {
```

Comment:
As an addition to my own comment.. `city` above is now a simple type (String) but what if it would be a bean itself (on both sides). Then it would again be logical to resume  name based mappings inside `city`.